src/app/queries/plans.ts

```diff
@@ -136,35 +140,111 @@ export const usePatchPlanMutation = (
+              `${forms.general.values.planName}-hook-`
+            ),
+            existingHook: true,
+            instance: instance,
+          };
+        }
+        return {
+          cr: generateHook(instance, null, `${forms.general.values.planName}-hook-`),
+          existingHook: false,
+          instance: instance,
+        };
+      });
+
+      const updatedHooksRef = planHooks.map(async (hook) => {
+        if (hook.existingHook) {
+          const response = await client.patch<IHook>(
+            hookResource,
+            (hook.cr.metadata as IMetaObjectMeta).name,
+            hook.cr
+          );
+          return { ref: nameAndNamespace(response.data.metadata), instance: hook.instance };
+        }
+        const response = await client.create<IHook>(hookResource, hook.cr);
+        return { ref: nameAndNamespace(response.data.metadata), instance: hook.instance };
+      });
+
+      const hooksRef = await Promise.all(updatedHooksRef);
+
+      const updatedPlan = generatePlan(forms, networkMappingRef, storageMappingRef, hooksRef);
       const [, , planResponse] = await Promise.all([
         networkMapping &&
           client.patch<Mapping>(networkMapResource, networkMappingRef.name, networkMapping),
         storageMapping &&
           client.patch<Mapping>(storageMapResource, storageMappingRef.name, storageMapping),
+
         client.patch<IPlan>(planResource, planBeingEdited.metadata.name, updatedPlan),
       ]);
 
-      // TODO handle hooks in the API
-      //  - for each instance in forms.hooks.values.instances:
-      //    - if there is a prefilledFromHook property in the instance, patch that existing hook using generateHook
-      //    - if not, create a new hook using generateHook
-      //  - for each hook reference in the plan (plan.spec.vms[].hooks[].hook) find any that don't match instances in the form data
-      //    - those were removed in the wizard, so delete the matching hook CRs
+      // Patch new hooks with ownerReferences to the edited plan
+      const hooksWithOwnerRef: IHook[] = [];
+      for (let i = 0; i < hooksRef.length; i++) {
+        if (!hooksRef[i].instance.prefilledFromHook) {
+          hooksWithOwnerRef.push(
+            generateHook(hooksRef[i].instance, hooksRef[i].ref, '', planBeingEdited)
+          );
+        }
+      }
+      const newHooksWithOwnerRef = hooksWithOwnerRef.map(async (hookWithOwnerRef) => {
+        const response = await client.patch<IHook>(
+          hookResource,
+          (hookWithOwnerRef.metadata as IMetaObjectMeta).name,
+          hookWithOwnerRef
+        );
+        return response;
+      });
+      await Promise.all(newHooksWithOwnerRef);
+
+      // Delete hooks removed from plan
+      const planHooksToDelete = planBeingEdited.spec.vms[0].hooks?.filter((vmsHook) => {
+        if (
+          forms.hooks.values.instances.length === 0 ||
+          forms.hooks.values.instances.filter((hook) => {
```

Comment:
Have you been able to test this? Reading through the code here it doesn't seem like this would work correctly. If there is at least one hook instance that was not prefilled, then `!hook.PrefilledFromHook` will be true and short circuit the name comparison, and the length of this filtered array will not be zero regardless of the value of `vmsHook`. So if you've created a new hook and deleted an old one, the old one's CR won't actually be deleted.

I think the logic you meant to use here was `if (hook.prefilledFromHook && ...)` instead of `if (!hook.prefilledFromHook || ...)`. But, I think you can simplify this entire block to:

```ts
      const planHooksToDelete = planBeingEdited.spec.vms[0].hooks?.filter(
        (vmsHook) =>
          !forms.hooks.values.instances.find((instance) =>
            isSameResource(instance.prefilledFromHook?.metadata, vmsHook.hook)
          )
      );
```

Reply:
For a given hook, 
if there are no hooks instances in form then the given hook is to be deleted (case all hooks are deleted from plan)
if there are hooks instances:
  then if the instance doesn't have prefilled data then it's a new hook
          or if it the instance has prefilled data and it's the same hook name then the hook is to be kept

I used that logic to create `planHooksToDelete`. 
The hooksToDelete is to map hooks (to delete) from the VM hooks (to delete) since they're different structures.

It works fine, I effectively tested it.

That said, your approach is neat (and makes me realize I'm still progressing :) ).
I'll use your then.
Thanks

Reply:
Maybe it's just an issue of code readability and I missed something then :) no problem.

I think it did work fine unless you had both created a new hook instance AND deleted a hook instance in the same plan edit operation. In that case, the `forms.hooks.values.instances.filter(...).length === 0` would never be true, because there is a new instance that causes `if (!hook.prefilledFromHook)` to be true and short circuit the `name === name` comparison in every iteration of the filter loop. So it doesn't matter which `vmsHook` we're filtering on, the logic says that none of them should be deleted as long as any non-prefilled instance is present. I may be misreading it though. Anyway, simplifying sounds good, thanks for taking my suggestion :)